quickwit/quickwit-proto/src/types/index_uid.rs

```diff
@@ -0,0 +1,208 @@
+impl ::prost::Message for IndexUid {
+    fn encode_raw<B>(&self, buf: &mut B)
+    where B: ::prost::bytes::BufMut {
+        if self.index_id != "" {
+            ::prost::encoding::string::encode(1u32, &self.index_id, buf);
+        }
+        let (ulid_high, ulid_low): (u64, u64) = self.incarnation_id.into();
+
+        if ulid_high != 0u64 {
+            ::prost::encoding::uint64::encode(2u32, &ulid_high, buf);
+        }
+        if ulid_low != 0u64 {
+            ::prost::encoding::uint64::encode(3u32, &ulid_low, buf);
+        }
+    }
+
+    fn merge_field<B>(
+        &mut self,
+        tag: u32,
+        wire_type: ::prost::encoding::WireType,
+        buf: &mut B,
+        ctx: ::prost::encoding::DecodeContext,
+    ) -> ::core::result::Result<(), ::prost::DecodeError>
+    where
+        B: ::prost::bytes::Buf,
+    {
+        const STRUCT_NAME: &'static str = "IndexUid";
+
+        match tag {
+            1u32 => {
+                let value = &mut self.index_id;
+                ::prost::encoding::string::merge(wire_type, value, buf, ctx).map_err(|mut error| {
+                    error.push(STRUCT_NAME, "index_id");
+                    error
+                })
+            }
+            2u32 => {
+                let (mut ulid_high, ulid_low) = self.incarnation_id.into();
+                ::prost::encoding::uint64::merge(wire_type, &mut ulid_high, buf, ctx).map_err(
+                    |mut error| {
+                        error.push(STRUCT_NAME, "incarnation_id_high");
+                        error
+                    },
+                )?;
+                self.incarnation_id = (ulid_high, ulid_low).into();
+                Ok(())
+            }
+            3u32 => {
+                let (ulid_high, mut ulid_low) = self.incarnation_id.into();
+                ::prost::encoding::uint64::merge(wire_type, &mut ulid_low, buf, ctx).map_err(
+                    |mut error| {
+                        error.push(STRUCT_NAME, "incarnation_id_low");
+                        error
+                    },
+                )?;
+                self.incarnation_id = (ulid_high, ulid_low).into();
+                Ok(())
+            }
+            _ => ::prost::encoding::skip_field(wire_type, tag, buf, ctx),
+        }
+    }
+
+    #[inline]
+    fn encoded_len(&self) -> usize {
+        let mut len = 0;
+
+        if self.index_id != "" {
+            len += ::prost::encoding::string::encoded_len(1u32, &self.index_id);
+        }
+        let (ulid_high, ulid_low): (u64, u64) = self.incarnation_id.into();
+
+        if ulid_high != 0u64 {
+            len += ::prost::encoding::uint64::encoded_len(2u32, &ulid_high);
+        }
+        if ulid_low != 0u64 {
+            len += ::prost::encoding::uint64::encoded_len(3u32, &ulid_low);
+        }
+        len
+    }
+
+    fn clear(&mut self) {
+        self.index_id.clear();
+        self.incarnation_id = Ulid::nil();
+    }
+}
+
+#[macro_export]
+macro_rules! index_uid {
```

Comment:
One way to move the `index_uid` field from an option without moving the other fields of the protobuf message.

Reply:
Trinity suggests using the getter and cloning if an owned object is required. The call to `clone()` will hopefully be optimized away. I agree :)